exec/cnex/cnex.c

```diff
@@ -842,19 +848,51 @@ void exec_INDEXAN(TExecutor *self)
     push(self->stack, val);
 }
 
-void exec_INDEXDR()
+void exec_INDEXDR(TExecutor *self)
 {
-    fatal_error("exec_INDEXDR not implemented");
+    self->ip++;
+    TString *index = string_fromString(top(self->stack)->string); pop(self->stack);
+    Cell *addr = top(self->stack)->address; pop(self->stack);
+    Cell *e = dictionary_findDictionaryEntry(addr->dictionary, index);
+    if (!e) {
+        char *pszIndex = string_asCString(index);
+        self->rtl_raise(self, "DictionaryIndexException", pszIndex, BID_ZERO);
+        free(pszIndex);
+        string_freeString(index);
+        return;
+    }
+    push(self->stack, cell_fromAddress(cell_dictionaryIndexForRead(addr, index)));
+    string_freeString(index);
 }
 
-void exec_INDEXDW()
+void exec_INDEXDW(TExecutor *self)
 {
-    fatal_error("exec_INDEXDW not implemented");
+    self->ip++;
+    TString *index = string_fromString(top(self->stack)->string); pop(self->stack);
+    Cell *addr = top(self->stack)->address; pop(self->stack);
+    push(self->stack, cell_fromAddress(cell_dictionaryIndexForWrite(addr, index)));
+//    string_freeString(index);
```

Comment:
Why is this commented out? Who is responsible for freeing the string? It looks like `cell_dictionaryIndexForWrite` doesn't, at least not in all cases.

Reply:
The string created by this call get passed to the dictionary object that is created, and becomes owned by the TDictionaryEntry object, and is therefore free()'ed when the Dictionary Object is freed.

Reply:
That's not the case if the dictionary key already exists in the dictionary. The cell_dictionaryIndexForWrite function only holds the index pointer if the element does *not* already exist in the dictionary.